File: sdk/runanywhere-flutter/lib/backends/onnx/onnx.dart
/// ONNX Runtime backend for RunAnywhere Flutter SDK.
///
/// This module provides STT, TTS, VAD, and LLM capabilities via the native
/// runanywhere-core library using Dart FFI.
///
/// ## Quick Start (matches iOS ONNX API exactly)
///
/// ```dart
/// import 'package:runanywhere/backends/onnx/onnx.dart';
///
/// // Register the module (matches iOS ONNX.register())
/// await Onnx.register();
///
/// // Add STT models
/// Onnx.addModel(
///   name: 'Sherpa Whisper Tiny (ONNX)',
///   url: 'https://github.com/.../sherpa-onnx-whisper-tiny.en.tar.gz',
///   modality: ModelCategory.speechRecognition,
///   memoryRequirement: 75000000,
/// );
///
/// // Add TTS models
/// Onnx.addModel(
///   name: 'Piper TTS (US English)',
///   url: 'https://github.com/.../vits-piper-en_US-lessac-medium.tar.gz',
///   modality: ModelCategory.speechSynthesis,
///   memoryRequirement: 65000000,
/// );
/// ```
///
/// ## What This Provides
///
/// - **STT (Speech-to-Text)**: Streaming and batch transcription
/// - **TTS (Text-to-Speech)**: Neural voice synthesis using VITS models
/// - **VAD (Voice Activity Detection)**: Real-time speech detection
/// - **LLM (Language Models)**: Text generation
library runanywhere_onnx;

import 'package:runanywhere/backends/onnx/onnx_download_strategy.dart';
import 'package:runanywhere/backends/onnx/providers/onnx_llm_provider.dart';
import 'package:runanywhere/backends/onnx/providers/onnx_stt_provider.dart';
import 'package:runanywhere/backends/onnx/providers/onnx_tts_provider.dart';
import 'package:runanywhere/backends/onnx/providers/onnx_vad_provider.dart';
import 'package:runanywhere/core/models/framework/model_artifact_type.dart';
import 'package:runanywhere/core/models/model/model_category.dart';
import 'package:runanywhere/core/models/model/model_info.dart';
import 'package:runanywhere/core/module/capability_type.dart';
import 'package:runanywhere/core/module/inference_framework.dart';
import 'package:runanywhere/core/module/model_storage_strategy.dart';
import 'package:runanywhere/core/module/runanywhere_module.dart';
import 'package:runanywhere/core/module_registry.dart';
import 'package:runanywhere/core/protocols/downloading/download_strategy.dart';
import 'package:runanywhere/foundation/logging/sdk_logger.dart';
import 'package:runanywhere/native/native_backend.dart';
import 'package:runanywhere/native/platform_loader.dart';

export 'onnx_download_strategy.dart';
export 'providers/onnx_llm_provider.dart';
export 'providers/onnx_stt_provider.dart';
export 'providers/onnx_tts_provider.dart';
export 'providers/onnx_vad_provider.dart';
export 'services/onnx_llm_service.dart';
export 'services/onnx_stt_service.dart';
export 'services/onnx_tts_service.dart';
export 'services/onnx_vad_service.dart';

// ============================================================================
// ONNX Module Implementation
// Matches iOS ONNX enum from ONNXRuntime/ONNXServiceProvider.swift
// ============================================================================

/// ONNX module for STT, TTS, VAD, and LLM capabilities.
///
/// Provides speech-to-text, text-to-speech, voice activity detection,
/// and language model services using ONNX Runtime with Sherpa-ONNX models.
///
/// Matches iOS `ONNX` enum from ONNXRuntime/ONNXServiceProvider.swift.
///
/// ## Registration (matches iOS ONNX pattern exactly)
///
/// ```dart
/// // Register module (matches iOS: ONNX.register())
/// await Onnx.register();
///
/// // Or with custom priority
/// await Onnx.register(priority: 150);
///
/// // Add models (matches iOS: ONNX.addModel())
/// Onnx.addModel(
///   name: 'Sherpa Whisper Tiny',
///   url: 'https://github.com/.../model.tar.gz',
///   modality: ModelCategory.speechRecognition,
/// );
/// ```
class Onnx extends RunAnywhereModule {
  // ============================================================================
  // Singleton Pattern (matches iOS enum pattern)
  // ============================================================================

  /// Singleton instance for module metadata
  static final Onnx _instance = Onnx._internal();

  /// Get the module instance (for use with ModuleRegistry.registerModule)
  static Onnx get module => _instance;

  Onnx._internal();

  // ============================================================================
  // Module State
  // ============================================================================

  static final SDKLogger _logger = SDKLogger(category: 'Onnx');
  static bool _isRegistered = false;
  static NativeBackend? _backend;
  static OnnxDownloadStrategy? _downloadStrategyInstance;

  // ============================================================================
  // RunAnywhereModule Implementation (matches iOS ONNX enum)
  // ============================================================================

  @override
  String get moduleId => 'onnx';

  @override
  String get moduleName => 'ONNX Runtime';

  @override
  InferenceFramework get inferenceFramework => InferenceFramework.onnx;

  @override
  Set<CapabilityType> get capabilities => {
        CapabilityType.stt,
        CapabilityType.tts,
        CapabilityType.vad,
        CapabilityType.llm,
      };

  @override
  int get defaultPriority => 100;

  @override
  ModelStorageStrategy? get storageStrategy => null;

  @override
  DownloadStrategy? get downloadStrategy {
    _downloadStrategyInstance ??= OnnxDownloadStrategy();
    return _downloadStrategyInstance;
  }

  // ============================================================================
  // Static API (matches iOS ONNX static methods exactly)
  // ============================================================================

  /// Whether the module is registered
  static bool get isRegistered => _isRegistered;

  /// Get native backend (for advanced usage)
  static NativeBackend? get backend => _backend;

  /// Check if the native backend is available on this platform.
  static bool get isAvailable {
    try {
      PlatformLoader.load();
      return true;
    } catch (_) {
      return false;
    }
  }

  /// Get the native library version.
  static String get version => _backend?.version ?? 'not initialized';

  /// Get backend info as a map.
  static Map<String, dynamic> getBackendInfo() {
    if (_backend == null) return {};
    return _backend!.getBackendInfo();
  }

  /// List of available backend names from the native library.
  static List<String> get availableBackends {
    if (_backend != null) {
      try {
        return _backend!.getAvailableBackends();
      } catch (_) {
        return [];
      }
    }
    if (!isAvailable) return [];
    NativeBackend? tempBackend;
    try {
      tempBackend = NativeBackend();
      return tempBackend.getAvailableBackends();
    } catch (_) {
      return [];
    } finally {
      tempBackend?.dispose();
    }
  }

  // ============================================================================
  // Registration (matches iOS ONNX.register() exactly)
  // ============================================================================

  /// Register ONNX module with the SDK.
  ///
  /// Matches iOS `ONNX.register(priority:)` pattern exactly.
  /// Registers STT, TTS, VAD, and LLM providers.
  ///
  /// [priority] - Registration priority (higher = preferred). Default: 100.
  static Future<void> register({int priority = 100}) async {
    if (_isRegistered) {
      _logger.debug('ONNX already registered');
      return;
    }

    // Check native library availability
    if (!isAvailable) {
      _logger.error('ONNX native library not available');
      return;
    }

    // Create native backend
    _backend = NativeBackend();
    _backend!.create('onnx');

    // Register as a module with ModuleRegistry (matches iOS pattern)
    ModuleRegistry.shared.registerModule(_instance, priority: priority);

    // Register all capability providers
    final registry = ModuleRegistry.shared;

    registry.registerSTT(
      OnnxSTTServiceProvider(_backend!),
      priority: priority,
    );

    registry.registerTTS(
      OnnxTTSServiceProvider(_backend!),
      priority: priority,
    );

    registry.registerVAD(
      OnnxVADServiceProvider(_backend!),
      priority: priority,
    );

    registry.registerLLM(
      OnnxLLMServiceProvider(_backend!),
      priority: priority,
    );

    _isRegistered = true;
    _logger
        .info('ONNX Runtime registered with capabilities: STT, TTS, VAD, LLM');
  }

  /// Register only ONNX STT service.
  ///
  /// Matches iOS `ONNX.registerSTT(priority:)` pattern.
  static Future<void> registerSTT({int priority = 100}) async {
    await _ensureBackendInitialized();

    ModuleRegistry.shared.registerSTT(
      OnnxSTTServiceProvider(_backend!),
      priority: priority,
    );
    _logger.info('ONNX STT registered');
  }

  /// Register only ONNX TTS service.
  ///
  /// Matches iOS `ONNX.registerTTS(priority:)` pattern.
  static Future<void> registerTTS({int priority = 100}) async {
    await _ensureBackendInitialized();

    ModuleRegistry.shared.registerTTS(
      OnnxTTSServiceProvider(_backend!),
      priority: priority,
    );
    _logger.info('ONNX TTS registered');
  }

  /// Register only ONNX VAD service.
  static Future<void> registerVAD({int priority = 100}) async {
    await _ensureBackendInitialized();

    ModuleRegistry.shared.registerVAD(
      OnnxVADServiceProvider(_backend!),
      priority: priority,
    );
    _logger.info('ONNX VAD registered');
  }

  /// Register only ONNX LLM service.
  static Future<void> registerLLM({int priority = 100}) async {
    await _ensureBackendInitialized();

    ModuleRegistry.shared.registerLLM(
      OnnxLLMServiceProvider(_backend!),
      priority: priority,
    );
    _logger.info('ONNX LLM registered');
  }

  /// Ensure native backend is initialized.
  static Future<void> _ensureBackendInitialized() async {
    if (_backend != null) return;

    if (!isAvailable) {
      throw StateError('ONNX native library not available');
    }

    _backend = NativeBackend();
    _backend!.create('onnx');
  }

  // ============================================================================
  // Model Registration (matches iOS ONNX.addModel() exactly)
  // ============================================================================

  /// Add a model to this module.
  ///
  /// Matches iOS `ONNX.addModel()` pattern exactly.
  /// Uses the module's inferenceFramework automatically.
  ///
  /// [id] - Explicit model ID. If null, generated from URL filename.
  /// [name] - Display name for the model.
  /// [url] - Download URL string for the model.
  /// [modality] - Model category (speechRecognition, speechSynthesis, etc.).
  /// [artifactType] - How the model is packaged (e.g., tarGzArchive).
  /// [memoryRequirement] - Estimated memory usage in bytes.
  /// [supportsThinking] - Whether the model supports reasoning/thinking.
  ///
  /// Returns the created ModelInfo, or null if URL is invalid.
  static ModelInfo? addModel({
    String? id,
    required String name,
    required String url,
    ModelCategory? modality,
    ModelArtifactType? artifactType,
    int? memoryRequirement,
    bool supportsThinking = false,
  }) {
    return _instance.addModelInternal(
      id: id,
      name: name,
      url: url,
      modality: modality,
      artifactType: artifactType,
      memoryRequirement: memoryRequirement,
      supportsThinking: supportsThinking,
    );
  }

  // ============================================================================
  // Cleanup
  // ============================================================================

  /// Dispose of resources (for cleanup)
  static void dispose() {
    if (_backend != null) {
      _backend!.dispose();
      _backend = null;
    }
    _isRegistered = false;
    _logger.info('ONNX disposed');
  }
}

File: sdk/runanywhere-react-native/src/Modules/ONNX.ts
/**
 * ONNX.ts
 *
 * ONNX Runtime module wrapper for RunAnywhere React Native SDK.
 * Provides public API for module registration and model declaration.
 *
 * This mirrors the Swift SDK's ONNX module pattern:
 * - ONNX.register() - Register the module with ServiceRegistry
 * - ONNX.addModel() - Declare a model for this module
 *
 * Reference: sdk/runanywhere-swift/Sources/ONNXRuntime/ONNXServiceProvider.swift
 *           sdk/runanywhere-swift/Sources/RunAnywhere/Core/Module/RunAnywhereModule.swift
 */

import { registerONNXProviders } from '../Providers/ONNXProvider';
import { ModelRegistry } from '../services/ModelRegistry';
import {
  LLMFramework,
  ModelCategory,
  ModelFormat,
  ConfigurationSource,
} from '../types/enums';
import type { ModelInfo } from '../types/models';
import { SDKLogger } from '../Foundation/Logging/Logger/SDKLogger';

const logger = new SDKLogger('ONNX');

/**
 * Model artifact type for ONNX models
 *
 * Matches iOS: ModelArtifactType enum
 */
export enum ModelArtifactType {
  /** Single file model */
  SingleFile = 'singleFile',
  /** Tar.gz archive with nested directory structure */
  TarGzArchive = 'tarGzArchive',
  /** Tar.bz2 archive with nested directory structure */
  TarBz2Archive = 'tarBz2Archive',
  /** ZIP archive */
  ZipArchive = 'zipArchive',
}

/**
 * Model registration options for ONNX models
 *
 * Matches iOS: ONNX.addModel() parameter structure
 */
export interface ONNXModelOptions {
  /** Unique model ID. If not provided, generated from URL filename */
  id?: string;
  /** Display name for the model */
  name: string;
  /** Download URL for the model */
  url: string;
  /** Model category (STT or TTS) */
  modality: ModelCategory;
  /** How the model is packaged (inferred from URL if not specified) */
  artifactType?: ModelArtifactType;
  /** Memory requirement in bytes */
  memoryRequirement?: number;
}

/**
 * ONNX Runtime Module
 *
 * Public API for registering ONNX module and declaring STT/TTS models.
 * This provides the same developer experience as the iOS SDK.
 *
 * ## Usage
 *
 * ```typescript
 * import { ONNX, ModelCategory } from 'runanywhere-react-native';
 *
 * // Register module
 * ONNX.register();
 *
 * // Add STT model
 * ONNX.addModel({
 *   id: 'sherpa-onnx-whisper-tiny.en',
 *   name: 'Sherpa Whisper Tiny (ONNX)',
 *   url: 'https://github.com/RunanywhereAI/sherpa-onnx/releases/download/runanywhere-models-v1/sherpa-onnx-whisper-tiny.en.tar.gz',
 *   modality: ModelCategory.SpeechRecognition,
 *   artifactType: ModelArtifactType.TarGzArchive,
 *   memoryRequirement: 75_000_000
 * });
 *
 * // Add TTS model
 * ONNX.addModel({
 *   id: 'vits-piper-en_US-lessac-medium',
 *   name: 'Piper TTS (US English - Medium)',
 *   url: 'https://github.com/RunanywhereAI/sherpa-onnx/releases/download/runanywhere-models-v1/vits-piper-en_US-lessac-medium.tar.gz',
 *   modality: ModelCategory.SpeechSynthesis,
 *   memoryRequirement: 65_000_000
 * });
 * ```
 *
 * Matches iOS: public enum ONNX: RunAnywhereModule
 */
export const ONNX = {
  /**
   * Module metadata
   * Matches iOS: static let moduleId, moduleName, inferenceFramework
   */
  moduleId: 'onnx',
  moduleName: 'ONNX Runtime',
  inferenceFramework: LLMFramework.ONNX,
  capabilities: ['stt', 'tts'] as const,
  defaultPriority: 100,

  /**
   * Register ONNX module with the SDK
   *
   * This registers both ONNX STT and TTS providers with ServiceRegistry,
   * enabling them to handle Sherpa-ONNX and Piper models.
   *
   * Matches iOS: static func register(priority: Int = defaultPriority)
   *
   * @example
   * ```typescript
   * ONNX.register();
   * ```
   */
  register(): void {
    logger.info('Registering ONNX module (STT + TTS)');
    registerONNXProviders();
    logger.info('✅ ONNX module registered (STT + TTS)');
  },

  /**
   * Add a model to this module
   *
   * Registers an ONNX model (STT or TTS) with the ModelRegistry.
   * The model will use ONNX framework automatically.
   *
   * Matches iOS: static func addModel(id:name:url:modality:artifactType:memoryRequirement:)
   *
   * @param options - Model registration options
   * @returns The created ModelInfo
   *
   * @example
   * ```typescript
   * // STT Model
   * ONNX.addModel({
   *   id: 'sherpa-onnx-whisper-small.en',
   *   name: 'Sherpa Whisper Small (ONNX)',
   *   url: 'https://github.com/k2-fsa/sherpa-onnx/releases/download/asr-models/sherpa-onnx-whisper-small.en.tar.bz2',
   *   modality: ModelCategory.SpeechRecognition,
   *   artifactType: ModelArtifactType.TarBz2Archive,
   *   memoryRequirement: 250_000_000
   * });
   *
   * // TTS Model
   * ONNX.addModel({
   *   id: 'vits-piper-en_GB-alba-medium',
   *   name: 'Piper TTS (British English)',
   *   url: 'https://github.com/RunanywhereAI/sherpa-onnx/releases/download/runanywhere-models-v1/vits-piper-en_GB-alba-medium.tar.gz',
   *   modality: ModelCategory.SpeechSynthesis,
   *   memoryRequirement: 65_000_000
   * });
   * ```
   */
  addModel(options: ONNXModelOptions): ModelInfo {
    // Generate stable ID from URL if not provided
    const modelId = options.id ?? this._generateModelId(options.url);

    // Infer artifact type from URL if not specified
    const artifactType =
      options.artifactType ?? this._inferArtifactType(options.url);

    // Format is always ONNX for this module
    const format = ModelFormat.ONNX;

    const now = new Date().toISOString();

    const modelInfo: ModelInfo = {
      id: modelId,
      name: options.name,
      category: options.modality,
      format,
      downloadURL: options.url,
      localPath: undefined,
      downloadSize: undefined,
      memoryRequired: options.memoryRequirement,
      compatibleFrameworks: [LLMFramework.ONNX],
      preferredFramework: LLMFramework.ONNX,
      supportsThinking: false, // ONNX STT/TTS models don't support thinking
      metadata: {
        tags: [],
        // artifactType stored separately, not in metadata
      },
      source: ConfigurationSource.Local,
      createdAt: now,
      updatedAt: now,
      syncPending: false,
      usageCount: 0,
      isDownloaded: false,
      isAvailable: true,
    };

    // Register with ModelRegistry (synchronous - queues async registration)
    ModelRegistry.registerModel(modelInfo);

    logger.info(`Added model: ${modelId} (${options.name})`);

    return modelInfo;
  },

  /**
   * Generate a stable model ID from URL
   * @internal
   */
  _generateModelId(url: string): string {
    try {
      const urlObj = new URL(url);
      const pathname = urlObj.pathname;
      const filename = pathname.split('/').pop() ?? 'model';
      // Remove common archive extensions
      return filename.replace(/\.(tar\.gz|tar\.bz2|zip|onnx)$/i, '');
    } catch {
      // Fallback for invalid URLs
      return `model-${Date.now()}`;
    }
  },

  /**
   * Infer artifact type from URL
   * @internal
   */
  _inferArtifactType(url: string): ModelArtifactType {
    const lowercased = url.toLowerCase();

    if (lowercased.includes('.tar.gz')) {
      return ModelArtifactType.TarGzArchive;
    } else if (lowercased.includes('.tar.bz2')) {
      return ModelArtifactType.TarBz2Archive;
    } else if (lowercased.includes('.zip')) {
      return ModelArtifactType.ZipArchive;
    }

    // Default to single file
    return ModelArtifactType.SingleFile;
  },
};
